### PR TITLE
Add custom status methods to RPC API

### PR DESCRIPTION
**Pull request by adriacidre**
This pull request exposes `status_login` and `status_signup` through a private ipc json rpc api.

Important changes:
- [x] Added new private IPC endpoint for `status_login`
- [x] Added new private IPC endpoint for `status_signup`
- [x] Endpoints will only be available through private calls through ISP when StatusServiceEnabled config variable is set.

This pull request is the first step of the [SDK swarm](https://ideas.status.im/ideas/146-status-go-sdk/README)

### CODE REVIEW

File: services/status/join_public_channel.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+)
+
+// JoinPublicStatusResponse : json response returned by status_joinpublicchannel.
+type JoinPublicStatusResponse struct {
+	Name     string `json:"name"`
+	Topic    string `json:"topic"`
+	FilterID string `json:"filterID"`
+}
+
+func joinPublicStatus(rpcClient *rpc.Client, rpcParams ...interface{}) (string, error) {
```
Comment (dshulyak) [inline services/status/join_public_channel.go]: you just need to generate symkey from password, and subscribe to filter? can you do it on client side?

File: services/status/api.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	api.whisper = w
+}
+
+// SetAccount sets account manager for the API calls.
+func (api *PublicAPI) SetAccount(a *account.Manager) {
+	api.am = a
+}
+
+// Login is an implementation of `status_login` or `web3.status.login` API
+func (api *PublicAPI) Login(context context.Context, rpcParams ...interface{}) (interface{}, error) {
```
Comment (dshulyak) [inline services/status/api.go]: why not normal objects?

File: geth/api/backend.go [hunk 3]
```diff
@@ -301,6 +318,10 @@ func (b *StatusBackend) registerHandlers() error {
 	rpcClient.RegisterHandler(params.PersonalSignMethodName, b.PersonalAPI().Sign)
 	rpcClient.RegisterHandler(params.PersonalRecoverMethodName, b.PersonalAPI().Recover)
 
+	rpcClient.RegisterHandler(params.StatusLoginMethodName, b.StatusAPI().Login)
```
Comment (dshulyak) [inline geth/api/backend.go]: why? you already registered an API

File: geth/api/backend_test.go [hunk 4]
```diff
@@ -129,6 +129,12 @@ func TestBackendGettersConcurrently(t *testing.T) {
 		wg.Done()
 	}()
 
+	wg.Add(1)
```
Comment (dshulyak) [inline geth/api/backend_test.go]: not sure this change belongs here

**Review by dshulyak — CHANGES_REQUESTED** [review]
i think there is some confusion about "normal" API and our handlers. if you registered a service with API you will  have `status_login` available (you probably tried it and wasn't able to access it, but that's because of `Public: false`, and you need explicitly whitelist that namespace).

Another thing is that i would prefer not to add joinchannel because it is a simple wrapper around whisper and can be done on client side, because you will consume messages from normal whisper API anyway, so no point to add an abstraction on that level.

handlers are used to inject our own logic for methods that already exist. for example, we don't want random dapps to access `personal_sign` because it will leak the password, thats why this handler was introduced so that any dapp can call this method, but then you need to confirm sign request using c-binding

this is the list of modules that are whitelsited https://github.com/status-im/status-go/pull/877/files#diff-41b02904df72dc739704e64f91cdc759R36

File: geth/api/backend.go [hunk 5]
```diff
@@ -142,6 +151,14 @@ func (b *StatusBackend) startNode(config *params.NodeConfig) (err error) {
...
 	b.transactor.SetRPC(b.statusNode.RPCClient(), rpc.DefaultCallTimeout)
 	b.personalAPI.SetRPC(b.statusNode.RPCPrivateClient(), rpc.DefaultCallTimeout)
+	b.statusAPI.SetRPC(b.statusNode.RPCPrivateClient(), rpc.DefaultCallTimeout)
+	b.statusAPI.SetAccount(b.AccountManager())
+	w, err := b.statusNode.WhisperService()
+	if err = b.registerHandlers(); err != nil {
+		b.log.Error("Error accessig whisper", "err", err)
+		return
+	}
+	b.statusAPI.SetWhisper(w)
```
Comment (dshulyak) [inline geth/api/backend.go]: can you do it same way it is done for shhext?

**Comment by gravityblast** [comment]
For functionality like joining a public chat, generating symmetric keys based on a password (chat name in the case of public chat), I think we can do everything in the `SDK` instead of doing it in the server. 
In the end, they are just `whisper` messages using the Status protocol so the `SDK` can just be a wrapper around the messages defined in the protocol.

About signup/login, do we really want our node to manage this?
A `Bot` only needs a key pair to sign messages and to receive contact requests. It can generate the keys using `shh_new_key_pair`, and save the id of the key. 

In general, I think it would be nice if we implement the `sdk` only as a wrapper to the Status protocol so that it can be used with any whisper node.

**Review by adambabik — COMMENTED** [review]
I am not sure this should be a part of the same stack as the rest of the APIs. It's a security concern and a new attack vector.

A few questions:
1. How will these methods be consumed? HTTP, IPC?
2. I though that things like "join public channel" will be a part of the SDK and will use public Whisper API. Has something changed?
3. I am not exactly sure how it is gonna work, but ideally, it would be great to have this RPC layer for bindings as a separate component that would not be compiled into mobile libraries.

**Comment by adriacidre** [comment]
Hey @adambabik,

*1.- How will these methods be consumed? HTTP, IPC?*
They should be consumed through json RPC calls

*2.- I though that things like "join public channel" will be a part of the SDK and will use public Whisper API. Has something changed?*
Yep, my bad, removing joining public channel ATM

*3.- I am not exactly sure how it is gonna work, but ideally, it would be great to have this RPC layer for bindings as a separate component that would not be compiled into mobile libraries.*
We can add a switch on the config file so the service only gets  active on demand.

**Comment by gravityblast** [comment]
> How will these methods be consumed? HTTP, IPC?

I think the `bot` should be able to specify if it wants to connect to a whisper node with `IPC` for a local node or `HTTP` for a local or remote node. Or maybe running `StatusNode` like the current bot example does, but there are other security issues related.

**Comment by adambabik** [comment]
With `HTTP` we need to be careful to avoid a few vector attacks:
- make sure CORS are properly configured to avoid calling it by malicious websites,
- DNS Rebinding,
- some more?

Thus, it would be able to use IPC.

**Comment by gravityblast** [comment]
@adambabik yeah maybe we can just start using IPC

**Comment by adriacidre** [comment]
@dshulyak all changes should be in place now

File: geth/node/node.go [hunk 6]
```diff
@@ -163,6 +170,27 @@ func activatePersonalService(stack *node.Node, config *params.NodeConfig) error
 	})
 }
 
+func activateStatusService(stack *node.Node, config *params.NodeConfig) error {
+	if !config.StatusServiceEnabled {
+		logger.Info("Status service api is disabled")
+		return nil
+	}
+
+	if config.WhisperConfig == nil || !config.WhisperConfig.Enabled {
```
Comment (adambabik) [inline geth/node/node.go]: I am not sure this is needed. If one wants to activate StatusService and Whisper is not found, it should throw an error.

File: services/status/api.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+// NewAPI creates an instance of the status API.
+func NewAPI(s *Service) *PublicAPI {
+	return &PublicAPI{s: s}
+}
+
+// LoginRequest : json request for status_login.
+type LoginRequest struct {
+	Addr string `json:"address"`
+	Pwd  string `json:"password"`
```
Comment (adambabik) [inline services/status/api.go]: `Password` is not much longer but it offers much better readability.

File: services/status/api.go [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+// LoginResponse : json response returned by status_login.
+type LoginResponse struct {
+	AccountKey   *keystore.Key `json:"-"`
+	AddressKeyID string        `json:"address_key_id"`
+}
+
+// Login is an implementation of `status_login` or `web3.status.login` API
+func (api *PublicAPI) Login(context context.Context, req LoginRequest) (LoginResponse, error) {
```
Comment (adambabik) [inline services/status/api.go]: Maybe use named return arguments `(response LoginResponse, err error)`?

File: services/status/api.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+// LoginRequest : json request for status_login.
+type LoginRequest struct {
+	Addr string `json:"address"`
+	Pwd  string `json:"password"`
+}
+
+// LoginResponse : json response returned by status_login.
+type LoginResponse struct {
+	AccountKey   *keystore.Key `json:"-"`
```
Comment (adambabik) [inline services/status/api.go]: Do we need to store it in this struct?

File: services/status/api.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+	Mnemonic string `json:"mnemonic"`
+}
+
+// Signup is an implementation of `status_signup` or `web3.status.signup` API
+func (api *PublicAPI) Signup(context context.Context, req SignupRequest) (SignupResponse, error) {
+	var err error
+	var res SignupResponse
+
+	if res.Address, res.Pubkey, res.Mnemonic, err = api.s.am.CreateAccount(req.Password); err != nil {
+		return res, ErrCouldNotCreateAnAccount
```
Comment (adambabik) [inline services/status/api.go]: The original `err` value is lost which might make it harder to debug what happened.

File: services/status/api.go [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+	"context"
+	"errors"
+
+	"github.com/ethereum/go-ethereum/accounts/keystore"
+)
+
+var (
+	// ErrInvalidStatusLoginParameters is returned when the number of parameters
+	// for status_login is not valid.
+	ErrInvalidStatusLoginParameters = errors.New("invalid number of parameters for status_login (2 expected)")
```
Comment (adambabik) [inline services/status/api.go]: Many of these errors are unused.

File: services/status/api.go [hunk 12]
```diff
@@ -0,0 +1,96 @@
+package status
```
Comment (adambabik) [inline services/status/api.go:1]: No unit tests?

File: t/e2e/services/generic_api_test.go [hunk 13]
```diff
@@ -0,0 +1,86 @@
+package services
```
Comment (adambabik) [inline t/e2e/services/generic_api_test.go]: This file only defines a suite but it does not contain any tests. Should it have `_test` prefix in the name?

File: t/e2e/services/generic_api_test.go [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+	"fmt"
+
+	"github.com/status-im/status-go/geth/params"
+	"github.com/status-im/status-go/sign"
+	"github.com/status-im/status-go/t/e2e"
+
+	. "github.com/status-im/status-go/t/utils"
+)
+
+type GenericRPCApiSuite struct {
```
Comment (adambabik) [inline t/e2e/services/generic_api_test.go]: `BaseJSONRPCSuite`?

File: t/e2e/services/generic_api_test.go [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+	e2e.BackendTestSuite
+}
+
+func (s *GenericRPCApiSuite) testAPIExported(method string, expectExported bool) {
```
Comment (adambabik) [inline t/e2e/services/generic_api_test.go]: It's a bit odd as it's named `testAPIExported` which suggest that it expects the API method to be available but it also takes `expectExported` param.

Maybe just create two methods: `AssertAPIMethodExported()` and `AssertAPIMethodUnexported()`?

File: t/e2e/services/generic_api_test.go [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+	s.NoError(json.Unmarshal([]byte(result), &response))
+
+	hidden := (response.Error != nil && response.Error.Code == methodNotFoundErrorCode)
+
+	s.Equal(expectExported, !hidden,
+		"method %s should be %s, but it isn't",
+		method, map[bool]string{true: "exported", false: "hidden"}[expectExported])
+}
+
+func (s *GenericRPCApiSuite) initTest(upstreamEnabled bool, statusServiceEnabled bool) error {
```
Comment (adambabik) [inline t/e2e/services/generic_api_test.go]: Why not `SetupTest()`? This is a standard method used by suites.

**Review by adambabik — COMMENTED** [review]
Also, it looks like many of tests put in `t/e2e` can be rewritten to unit tests.

**Review by dshulyak — APPROVED** [review]
definitely +1 for moving e2e tests to unit

File: t/e2e/services/generic_api_test.go (hunk 13 above)
Comment (adriacidre) [reply t/e2e/services/generic_api_test.go]: I can't find any example on the current test suite, you mean a **file name** like `_test_basic_api_test.go`?

Or this review comment is about package name so it's moved to something like `services_test`?

**Comment by adriacidre** [comment]
@adambabik all changes should be addressed now except:
> This file only defines a suite but it does not contain any tests. Should it have _test prefix in the name?

please let me know how the filename should look like and i'll change it :)

**Comment by gravityblast** [comment]
@adriacidre @adambabik I'm still not sure if we really need the signup/login API in `statusd`.
The `signup` just creates a key pair and save it to the key store, but users should be able to use their own key pairs if they want, maybe generated in other ways.

In this example the code doesn't rely on `statusd` but still uses the protocol to interact with Status:

https://github.com/status-im/status-go-sdk/commit/84963606d7b951ed44772ff4d565dea4ebff0116

If we really want to have a simple way to generate the keys maybe it can be in the SDK, but I don't think  it belongs here.

What do you think?

**Comment by adriacidre** [comment]
hey @pilu, 

Main idea behind SDK is to attract new community GO devs to join and build new features not only on top of the SDK but also on status-go.

Can we just move forward with a status-go based backend for now? That will help us to inherit any new features built on top of status-go by reducing development efforts on SDK part.

**Comment by gravityblast** [comment]
@adriacidre yes I understand the idea, and I agree it would be easier from the SDK part.
So are we going to use this api also from `status-react` instead of calling `account.CreateAccount`?
Maybe it could make sense so that all the calls to create an account pass to the same part of the code.

`status-react` uses other APIs too, like `shh_post` to send the message. 
Would you like to use them as well from the sdk or do we want to have everything unified in the `status_` namespace, having something like `status_post`?

**Comment by adriacidre** [comment]
@pilu 
1.- No changes planned on status-react by now
2.- We're actually using a bunch of endpoints on the current version of the SDK `shh_generateSymKeyFromPassword`, `shh_newMessageFilter`, `shh_getFilterMessages`, `shh_post`, `web3_sha3` they're defined by now on [here](https://github.com/status-im/status-go-sdk/blob/master/dictionary.go#L3)

**Review by gravityblast — APPROVED** [review]
👍  🚀

File: services/status/api.go [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+// SignupResponse : json response returned by status_signup.
+type SignupResponse struct {
+	Address  string `json:"address"`
+	Pubkey   string `json:"pubkey"`
+	Mnemonic string `json:"mnemonic"`
+}
+
+// Signup is an implementation of `status_signup` or `web3.status.signup` API
+func (api *PublicAPI) Signup(context context.Context, req SignupRequest) (SignupResponse, error) {
+	var err error
```
Comment (adambabik) [inline services/status/api.go]: These vars can also be used as named arguments.

File: t/e2e/services/base_api_test.go [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+	. "github.com/status-im/status-go/t/utils"
+)
+
+type BaseJSONRPCSuite struct {
+	e2e.BackendTestSuite
+}
+
+func (s *BaseJSONRPCSuite) AssertAPIMethodUnexported(method string) {
+	exported := s.isMethodExported(method, false)
+	s.Equal(false, exported,
```
Comment (adambabik) [inline t/e2e/services/base_api_test.go]: It's fine but just a friendly reminder that there is `s.False/True()`,

File: t/e2e/services/base_api_test.go [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+	return s.Backend.StartNode(nodeConfig)
+}
+
+func (s *BaseJSONRPCSuite) notificationHandlerSuccess(account string, pass string) func(string) {
+	return func(jsonEvent string) {
+		s.notificationHandler(account, pass, nil)(jsonEvent)
+	}
+}
+
+func (s *BaseJSONRPCSuite) notificationHandler(account string, pass string, expectedError error) func(string) {
```
Comment (adambabik) [inline t/e2e/services/base_api_test.go]: I am not sure this method should be here. It looks like it belongs to Personal API suite.

**Review by adambabik — APPROVED** [review]
A great first step towards the Go SDK!

File: t/e2e/services/base_api_test.go (hunk 19 above)
Comment (adriacidre) [reply t/e2e/services/base_api_test.go]: cheers mate, all fixed now :)

**Pull request merged**